### PR TITLE
Table provider error propagation

### PR DESCRIPTION
**Pull request by jdye64**
Closes: #2436

### CODE REVIEW

File: datafusion/core/src/execution/context.rs [hunk 1]
```diff
@@ -1364,10 +1364,16 @@ impl SessionState {
...
-    fn get_table_provider(&self, name: TableReference) -> Option<Arc<dyn TableProvider>> {
+    fn get_table_provider(&self, name: TableReference) -> Result<Arc<dyn TableProvider>> {
         let resolved_ref = self.resolve_table_ref(name);
-        let schema = self.schema_for_ref(resolved_ref).ok()?;
-        schema.table(resolved_ref.table)
+        let schema = self.schema_for_ref(resolved_ref).unwrap();
+        match schema.table(resolved_ref.table) {
+            Some(e) => Ok(e),
+            None => Err(DataFusionError::Plan(format!(
+                "Table or CTE with name '{}' not found",
```
Comment (andygrove) [inline datafusion/core/src/execution/context.rs:1373]: We shouldn't be attempting to resolve CTEs via `get_table_provider`

```suggestion
                "Table with name '{}' not found",
```

File: datafusion/core/src/execution/context.rs [hunk 2]
```diff
@@ -1364,10 +1364,16 @@ impl SessionState {
...
-        let schema = self.schema_for_ref(resolved_ref).ok()?;
-        schema.table(resolved_ref.table)
+        let schema = self.schema_for_ref(resolved_ref).unwrap();
+        match schema.table(resolved_ref.table) {
+            Some(e) => Ok(e),
+            None => Err(DataFusionError::Plan(format!(
+                "Table with name '{}' not found",
+                name.table()
+            ))),
+        }
```
Comment (andygrove) [inline datafusion/core/src/execution/context.rs:1376]: nit: We could make this more concise using `ok_or_else`

```suggestion
        schema.table(resolved_ref.table).ok_or_else(|| {
            DataFusionError::Plan(format!("Table with name '{}' not found", name.table()))
        })
```

**Review by andygrove — APPROVED** [review]
LGTM. I left one code style suggestion

File: datafusion/core/src/execution/context.rs (hunk 2 above)
Comment (jdye64) [reply datafusion/core/src/execution/context.rs]: Yes, this is a much better suggestion.

File: datafusion/core/src/execution/context.rs [hunk 3]
```diff
@@ -1364,10 +1363,20 @@ impl SessionState {
...
 impl ContextProvider for SessionState {
-    fn get_table_provider(&self, name: TableReference) -> Option<Arc<dyn TableProvider>> {
+    fn get_table_provider(&self, name: TableReference) -> Result<Arc<dyn TableProvider>> {
         let resolved_ref = self.resolve_table_ref(name);
-        let schema = self.schema_for_ref(resolved_ref).ok()?;
-        schema.table(resolved_ref.table)
+        match self.schema_for_ref(resolved_ref) {
+            Ok(schema) => {
+                schema.table(resolved_ref.table).ok_or_else(|| {
+                    //DataFusionError::Plan(format!("Table with name '{}' not found", name.table()))
```
Comment (andygrove) [inline datafusion/core/src/execution/context.rs:1371]: nit: the commented code can be removed now

**Review by andygrove — APPROVED** [review]

**Comment by andygrove** [comment]
@jdye64 Looks like you need to run `cargo fmt` again

**Pull request merged**